src/compiler/compile/nodes/shared/Context.ts

```diff
@@ -44,13 +52,55 @@ export function unpack_destructuring(contexts: Context[], node: Node, modifier:
 				const key = property.key as Identifier;
 				const value = property.value;
 
-				used_properties.push(x`"${(key as Identifier).name}"`);
+				used_properties.push(x`"${key.name}"`);
 				if (value.type === 'AssignmentPattern') {
-					unpack_destructuring(contexts, value.left, node => x`${modifier(node)}.${key.name} !== undefined ? ${modifier(node)}.${key.name} : ${value.right}` as Node);
+					const n = contexts.length;
+
+					unpack_destructuring(contexts, value.left, node => {
+						const alternate = JSON.parse(JSON.stringify(value.right));
+
+						update_reference(contexts.slice(0, n), node, alternate);
+
+						return x`${modifier(node)}.${key.name} !== undefined ? ${modifier(node)}.${key.name} : ${alternate.replacement || alternate}` as Node;
+					});
 				} else {
 					unpack_destructuring(contexts, value, node => x`${modifier(node)}.${key.name}` as Node);
 				}
 			}
 		});
 	}
 }
+
+function update_reference(contexts: Context[], node: Node, parent: any, property?: any) {
+	const current_node = (property !== undefined && parent[property] || parent);
+
+	if (!current_node || !contexts.length) return;
+
+	if (current_node.type === 'Identifier') {
```

Comment:
probably can use `estree-walker` to traverse down the node.

Reply:
you can use `is-reference` to determine whether the Identifier is a reference or not.

Reply:
will try to use `estree-walker`, I've seen it to be used in some codes there but hesitated since doing recursive looks more compelling...

Reply:
done